### ISSUE TITLE
I cannot reproduce the small and medium size GPT-2 results.

### ISSUE THREAD

**Issue by dahunshin** [Liuhong99/Sophia#46]
Dear Authors

I reproduced the results of GPT-2 125M following your instructions(dependencies, config, commands), but they did not match the results in your paper. For example, in the validation loss vs. number of steps figure in your paper, Sophia-G outperforms AdamW on validation loss across all steps up to 100k steps. However, the results from my reproduction show that AdamW outperforms Sophia-G on validation loss during the middle of pretraining, from 20k to 80k steps.

Purple: Sophia-G
Yellow: AdamW

![gpt2-small](https://github.com/Liuhong99/Sophia/assets/62804637/e5d2d7b8-1f5f-47ef-a78c-3ebb8e7a30c6)

In addition, to see that Sophia is two times faster than AdamW, I pre-trained the model using AdamW for 200K steps and compared its validation loss to that of Sophia-G at 100K step. However, contrary to your claim, AdamW-200K achieved lower(better) validation loss than Sophia-G-100K, rather than achieving the same loss (AdamW-200K: 2.863 validation loss > Sophia-G-100K: 2.897).

![200k](https://github.com/Liuhong99/Sophia/assets/62804637/d5127af4-d848-422a-b3fe-1c38c72d8406)

Can you tell me how to get the same results in your papers? I hope you let me know there are things if I missed for the reproduction. 

Sincerely,
Dahun Shin

**Comment by asaiacai** [comment]
I also followed the instructions in the README and my validation loss curves are basically identical between AdamW and Sophia.

![W B Chart 11_10_2023, 1_17_52 AM](https://github.com/Liuhong99/Sophia/assets/25255966/52163641-3910-4c75-8ed4-76e243d70dbd)

![Screenshot 2023-11-10 at 1 19 49 AM](https://github.com/Liuhong99/Sophia/assets/25255966/2afd149c-1c56-41b3-b509-26bf57705257)

**Comment by zimingyy** [comment]
> Dear Authors
> 
> I reproduced the results of GPT-2 125M following your instructions(dependencies, config, commands), but they did not match the results in your paper. For example, in the validation loss vs. number of steps figure in your paper, Sophia-G outperforms AdamW on validation loss across all steps up to 100k steps. However, the results from my reproduction show that AdamW outperforms Sophia-G on validation loss during the middle of pretraining, from 20k to 80k steps.
> 
> Purple: Sophia-G Yellow: AdamW
> 
> ![gpt2-small](https://user-images.githubusercontent.com/62804637/280199259-e5d2d7b8-1f5f-47ef-a78c-3ebb8e7a30c6.png)
> 
> In addition, to see that Sophia is two times faster than AdamW, I pre-trained the model using AdamW for 200K steps and compared its validation loss to that of Sophia-G at 100K step. However, contrary to your claim, AdamW-200K achieved lower(better) validation loss than Sophia-G-100K, rather than achieving the same loss (AdamW-200K: 2.863 validation loss > Sophia-G-100K: 2.897).
> 
> ![200k](https://user-images.githubusercontent.com/62804637/280200775-d5127af4-d848-422a-b3fe-1c38c72d8406.png)
> 
> Can you tell me how to get the same results in your papers? I hope you let me know there are things if I missed for the reproduction.
> 
> Sincerely, Dahun Shin

Hi, I am similarly unable to reproduce Sophia's experiments, which differ significantly from the results in the paper, can I ask about your experimental setting? How many gpu were used and what type of gpu were used?

ALSO, are the hyperparameters you're using consistent with the original article?

**Comment by Liuhong99** [comment]
Hi @dahunshin @asaiacai and @Ziiiirem,
<img width="660" alt="image" src="https://github.com/Liuhong99/Sophia/assets/39693953/38b1cf39-e3cd-47cf-9d0a-52eb47e890dc">
https://api.wandb.ai/links/hliu99/68v9uc67
This is what I got with
```
batch_size=8
beta1=0.965
beta2=0.99
dtype="bfloat16"
rho=0.05
gradient_accumulation_steps=6
learning_rate=0.0006
max_iters=100000
min_lr=0.00001
```
with 10 A5000 GPUs. This should be 2x faster than AdamW baseline 200k run. I can help if you can share your configurations.

**Comment by dahunshin** [comment]
Thank you for the reply. 

Despite your presentation of a new figure, I remain unclear about your intended message. The validation loss I obtained for adamW-200k is 2.863, while your result for sophiaG-100k is only 2.875 (at 95k) and this is not the validation loss at the 100k iteration. According to your new figure, it appears that SophiaG's validation loss is worse at the 100k iteration.

Moreover, I have one more question: I cannot replicate the medium size GPT-2 as well.

I used 'train_gpt2_medium_sophiag.py' and 'train_gpt2_medium_adam.py' and A100(80GB) x 8 

The result I got is:

Green: adamW
Yellow: sophiaG

![mediumsize](https://github.com/Liuhong99/Sophia/assets/62804637/b99e7d6e-bff1-4f15-b081-405e1a4ae2ce)

This is different from the figure:

![author](https://github.com/Liuhong99/Sophia/assets/62804637/1ed15f35-94aa-4e5d-aefc-a8347df5abb2)

Again, could you let me know the hyper-parameters for the medium size GPT-2?

**Comment by Liuhong99** [comment]
Hi @dahunshin,

I have run the experiments for small and medium again. The hyper-parameters were updated in the config files. For small, the hyper parameters remain unchanged. For medium the lr is increased to 7e-4. 
<img width="1606" alt="Screenshot 2024-01-29 at 10 01 15 PM" src="https://github.com/Liuhong99/Sophia/assets/39693953/c082258d-e46e-4af1-8e14-9a7f0fe77c2f">
<img width="1603" alt="Screenshot 2024-01-29 at 10 01 34 PM" src="https://github.com/Liuhong99/Sophia/assets/39693953/59416c24-4677-440f-9668-323edf737a43">
<img width="1613" alt="Screenshot 2024-01-29 at 10 05 40 PM" src="https://github.com/Liuhong99/Sophia/assets/39693953/dd9b4b1b-f71c-4b79-a03a-6255b9b343da">

**Comment by Liuhong99** [comment]
we also need to set `scale_attn_by_inverse_layer_idx` to `True`

**Comment by dahunshin** [comment]
Your response is very helpful. Thank you.
